[docs] EKS & AKS auto-discovery docs (#17628)
gravitational/teleport: docs/pages/kubernetes-access/guides/discovery.mdx

---

> title: Kubernetes Discovery Guides (Preview)

Reviewer: ```suggestion
title: Kubernetes Discovery (Preview)
```
Suggesting naming the page after the feature, rather than the fact that it contains guides.

> description: Detailed guides for configuring Kubernetes Auto-discovery.

Reviewer: ```suggestion
description: Detailed guides for configuring Kubernetes Auto-Discovery.
```

> description: Detailed guides for configuring Kubernetes Auto-discovery.

Reviewer: Should we say "Kubernetes Auto-Discovery" everywhere, rather than "Kubernetes Discovery"?

layout: tocless-doc
---

> ## Discovery

Reviewer: ```suggestion
```
I don't think we need to title the introductory section.

Kubernetes cluster auto-discovery is a Teleport feature that allows the Teleport process automatically discover and enroll Kubernetes clusters from different cloud providers. 

> The discovery process does not install any component on the cluster but requires Teleport to have access to the cluster API and minimal permissions.

Reviewer: ```suggestion
The discovery process does not install any component on the cluster but requires Teleport to have access to a cluster's API and minimal permissions.
```

> 

Reviewer: ```suggestion
## How Kubernetes cluster auto-discovery works
```

The discovery consists of two distinct steps:

> ### Discovery

Reviewer: ```suggestion
### Polling cloud APIs
```
Or something else that's more descriptive. Since "discovery" is part of the name of the feature, I thought it would make this clearer to use a different term for the first step.

> The discovery process is responsible for scanning the cloud providers and checking if any Kubernetes cluster match the desired filtering labels. 

Reviewer: ```suggestion
The discovery process is responsible for scanning cloud providers and checking if any Kubernetes clusters match the desired filtering labels. 
```

> When a new cluster is discovered, it registers that cluster in Teleport together with the cloud access parameters.

Reviewer: Does this mean that the cluster registers itself with Teleport? Or does Teleport do the registering? Also, what is an example of a cloud access parameter?

> This process is also responsible for removing the cluster when it has been deleted or when it no longer meets the filtering labels.

Reviewer: ```suggestion
This process is also responsible for removing the cluster from Teleport's registry when it has been deleted or when it no longer meets the filtering labels.
```
Is this accurate?

> ```yaml

Reviewer: We should add a sentence to introduce this configuration snippet.

> # This section configures the 'discovery service'

Reviewer: If the discovery service is a product like the Teleport SSH/Database/Kubernetes etc. Service is, we should capitalize it (Discovery Service).

       tags:
         "*": "*"
    # Matchers for discovering Azure-hosted resources.
    azure:
      # Azure resource types. Valid options are:
      # 'aks' - discovers and registers Azure AKS Kubernetes Clusters.
    - types: ["aks"]
      # Azure regions to search for resources from. Valid options are:
      # '*' - discovers resources in all regions (default).
      # Any valid Azure region name. List all valid regions using the Azure "az" cli: `az account list-locations -o table`
      regions: ["*"]
      # Azure subscription IDs to search resources from. Valid options are:
      # '*' - discovers resources in all subscriptions (default).
      subscriptions: ["*"]
      # Azure resource groups to search resources from. Valid options are:
      # '*' - discovers resources in all resource groups within configured subscription(s) (default).
      resource_groups: ["*"]
      # Azure resource tag filters used to match resources.
      tags:
        "*": "*"
```

The `discovery_service` automatically imports the tags defined in the cluster and 
their value is constantly monitored and updated.

### Registering and forwarding requests

> The `kubernetes_service` is responsible for monitoring the resources created or 

Reviewer: ```suggestion
The Teleport Kubernetes Service is responsible for monitoring the resources created or 
```
Referring to the service name and not its configuration block

> updated by the `discovery_service` and forwarding requests to them.

Reviewer: ```suggestion
updated by the Discovery Service and forwarding requests to them.
```

> It requires direct access to the Cluster API and that the permissions associated 

Reviewer: ```suggestion
It requires direct access to the API of a target Kubernetes cluster and that the permissions associated 
```

> with the identity it is using are able to use `Impersonation`.

Reviewer: What does `Impersonation` refer to? Are we talking about using impersonation headers in Kubernetes?

> To watch and serve discovered clusters the following confuguration is required:

Reviewer: What does "serve" mean here?

```yaml
## This section configures the 'kubernetes service'
kubernetes_service:
    enabled: "yes"

>     # Matchers for dynamic kubernetes cluster resources created with "tctl create" command or by Kubernetes auto-discovery.

Reviewer: ```suggestion
    # Matchers for dynamic Kubernetes cluster resources created with the "tctl create" command or by Kubernetes auto-discovery.
```

    # When resources were created by 'discovery_service', ' kubernetes_service' must have the required permissions.
    resources:
    - labels:
        "*": "*" # can be configured to limit the clusters to watch for this service.
```

> Both services can be configured in the same Teleport process or in separate processes. 

Reviewer: ```suggestion
Both the Kubernetes Service and the Discovery Service can be configured in the same Teleport process or in separate processes. 
```

When they run separately, you need to take care of the permissions required for each service.

> ## Supported Clouds

Reviewer: ```suggestion
## Supported clouds
```
